### ISSUE TITLE
Deployment controller fails to create new RS when old RS has an evicted pod

### ISSUE THREAD

**Issue by KnVerey** [kubernetes/kubernetes#60162]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

The deployment controller for a `strategy: Recreate` deployment with a single replica failed to create a new ReplicaSet when the old ReplicaSet included an Evicted pod. Typically, the eviction was caused by "node low on resource: nodefs". When we manually deleted the evicted pod, the new RS and accompanying pod were created immediately. We have seen this happen at least a dozen times. Here are the audit logs for a recent case:

![image](https://user-images.githubusercontent.com/4789493/36509901-26d93d38-172f-11e8-9062-084bfc202afa.png)

**What you expected to happen**:

The deployment controller ignores the evicted pod and creates the updated replicaset/pod.

**How to reproduce it (as minimally and precisely as possible)**:

To reproduce, you would need a way to trigger a kubelet-driven eviction on a node that is currently running a single-replica Deployment with  `strategy: Recreate`. Wait for the pod to be evicted and the new pod to be created, confirm that the evicted pod still exists, and then make a trivial change to the deployment spec. The running pod will be deleted, and no new replicaset/pod will be created. 

This issue does not seem to occur for manual evictions (via `POST /eviction`); pods I evict that way disappear from the API server immediately.

**Anything else we need to know?**:

Possibly related to https://github.com/kubernetes/kubernetes/issues/54525 and/or https://github.com/kubernetes/kubernetes/issues/55630

**Environment**:
- Kubernetes version (use `kubectl version`):  v1.8.5-gke.0
- Cloud provider or hardware configuration: GKE
- OS (e.g. from /etc/os-release): Container-Optimized OS

**Comment by KnVerey** [comment]
@kubernetes/sig-apps-bugs

**Comment by tnozicka** [comment]
@janetkuo I am actually debugging the likely the same production issue with Recreate strategy in two clusters right now. (We run on patched 1.9.1.) I have traced it down to broken RS as well.

in 1 cluster RS leaves around pod failed matching the selector 
```
status:
  message: Pod Predicate MatchNodeSelector failed
  phase: Failed
  reason: MatchNodeSelector
  startTime: 2018-02-20T02:43:02Z
```
in the other one those are evicted pods. all replicas are scaled down to 0 and report 0 in status but pods with valid controllerRefs to them are still present.

> 1. A Recreate Deployment was updated by the user, so it scaled all old ReplicaSets to 0

Same

> 2. The old ReplicaSet was scaled to 0 (ReplicaSet's .spec.replicas == 0), but still owned the evicted pod, which wouldn't be killed automatically because of the above comment (ReplicaSet's .status.replicas > 0)

replicaSet.status.replicas is 0 for me but pod is still there

> We can either modify Deployment controller (to ignore evicted old pods) or ReplicaSet controller (to kill evicted pods when it gets scaled down), if this is not the intended behavior. 

I think the correct behaviour is for replicaSet scaled to 0 to delete all of its pod, not to leave them laying around

**Comment by tnozicka** [comment]
xref - this should fix it https://github.com/kubernetes/kubernetes/pull/60301
